core/src/main/java/org/mule/runtime/core/exception/TypedException.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.core.exception;
+
+import static org.mule.runtime.api.util.Preconditions.checkArgument;
+
+import org.mule.runtime.api.message.Error;
+import org.mule.runtime.api.message.ErrorType;
+
+/**
+ * Internal {@link RuntimeException} implementation to throw {@link Throwable throwables} that indicates explicitly
+ * the {@link ErrorType} that is wanted to throw.
+ * Also gives the possibility to declare to add a message and detailed message for the {@link Error} to built.
+ *
+ * @since 1.0
```

Comment:
Eventually we might need it there along with the ErrorMessageAware one, at least an interface that provides the error part? If it's not necessary now I wouldn't move it.